File: LeetCode/python3/986.py
class Solution:
    def intervalIntersection(self, A: 'List[Interval]', B: 'List[Interval]') -> 'List[Interval]':
        ret = []
        i, j, m, n = 0, 0, len(A), len(B)
        while i < m and j < n:
            if A[i].start > B[j].end:
                j += 1
            elif A[i].end < B[j].start:
                i += 1
            else:
                ret.append(Interval(max(A[i].start, B[j].start), min(A[i].end, B[j].end)))
                if A[i].end < B[j].end:
                    i += 1
                else:
                    j += 1
        return ret

File: LeetCode/cpp/986.cpp
class Solution {
public:
    vector<Interval> intervalIntersection(vector<Interval> &A, vector<Interval> &B) {
        int i = 0, j = 0, m = A.size(), n = B.size();
        if (m == 0 || n == 0)
            return {};
        vector<Interval> ret;
        while (i < m && j < n) {
            if (B[j].start > A[i].end)
                ++i;
            else if (A[i].start > B[j].end)
                ++j;
            else {
                ret.push_back(Interval(max(A[i].start, B[j].start), min(A[i].end, B[j].end)));
                if (A[i].end > B[j].end)
                    ++j;
                else
                    ++i;
            }
        }
        return ret;
    }
};